vispy/visuals/_scalable_textures.py

```diff
@@ -324,6 +324,8 @@ class GPUScaledTextureMixin(_ScaledTextureMixin):
         np.float64: 'r32f',
         np.uint8: 'r8',  # uint8 normalized
         np.uint16: 'r16',  # uint16 normalized
+        np.complex64: 'r32f',
+        np.complex128: 'r32f',
```

Comment:
Based on some feedback from @almarklein on my PR that added this functionality, I'm tempted to say these should be moved to the ComplexImageVisual. You can pass `r32f` directly as `texture_format` to `ImageVisual.__init__` and it should "just work". Having this in these base classes should include all the custom functionality for converting the numpy arrays to the proper storage solution. And if we're doing that then we might as well have a separate implementation of the Mixin class specifically for complex data...which is just complicated and is maybe easier to just put stuff in the ComplexImageVisual.

However, if you are planning on doing a ComplexVolumeVisual then maybe a separate base class for these mixins does make sense. But maybe that can wait until that PR.

Reply:
sure, I'll give that a shot.  volume visual probably won't be very soon.

Reply:
Unfortunately, it's not as easy as just passing "r32f" or "rg32f" to the `ImageVisual.__init__`.   This is what I get if I change "auto" to "r32f":
```python
Traceback (most recent call last):
  File "/Users/talley/Dropbox (HMS)/Python/forks/vispy/examples/basics/scene/complex_image.py", line 37, in <module>
    image = scene.visuals.ComplexImage(img_data, parent=view.scene, complex_mode="phase")
  File "/Users/talley/Dropbox (HMS)/Python/forks/vispy/vispy/scene/visuals.py", line 131, in __init__
    subclass.__init__(self, *args, **kwargs)
  File "/Users/talley/Dropbox (HMS)/Python/forks/vispy/vispy/visuals/image_complex.py", line 48, in __init__
    super().__init__(data=data, **kwargs)
  File "/Users/talley/Dropbox (HMS)/Python/forks/vispy/vispy/visuals/image.py", line 291, in __init__
    self.set_data(data)
  File "/Users/talley/Dropbox (HMS)/Python/forks/vispy/vispy/visuals/image_complex.py", line 62, in set_data
    return super().set_data(data)
  File "/Users/talley/Dropbox (HMS)/Python/forks/vispy/vispy/visuals/image.py", line 344, in set_data
    self._texture.check_data_format(data)
  File "/Users/talley/Dropbox (HMS)/Python/forks/vispy/vispy/visuals/_scalable_textures.py", line 429, in check_data_format
    raise ValueError("Data being set would cause a format change "
ValueError: Data being set would cause a format change in the texture. This is only allowed when 'texture_format' is set to 'auto'.

```

and this is what I get i get if I use 'rg32f' instead:

```python
Traceback (most recent call last):
  File "/Users/talley/Dropbox (HMS)/Python/forks/vispy/examples/basics/scene/complex_image.py", line 37, in <module>
    image = scene.visuals.ComplexImage(img_data, parent=view.scene, complex_mode="phase")
  File "/Users/talley/Dropbox (HMS)/Python/forks/vispy/vispy/scene/visuals.py", line 131, in __init__
    subclass.__init__(self, *args, **kwargs)
  File "/Users/talley/Dropbox (HMS)/Python/forks/vispy/vispy/visuals/image_complex.py", line 48, in __init__
    super().__init__(data=data, **kwargs)
  File "/Users/talley/Dropbox (HMS)/Python/forks/vispy/vispy/visuals/image.py", line 266, in __init__
    self._texture = self._init_texture(data, texture_format)
  File "/Users/talley/Dropbox (HMS)/Python/forks/vispy/vispy/visuals/image.py", line 321, in _init_texture
    tex = GPUScaledTexture2D(
  File "/Users/talley/Dropbox (HMS)/Python/forks/vispy/vispy/visuals/_scalable_textures.py", line 73, in __init__
    super().__init__(data, **texture_kwargs)
  File "/Users/talley/Dropbox (HMS)/Python/forks/vispy/vispy/gloo/texture.py", line 512, in __init__
    BaseTexture.__init__(self, data, format, resizable, interpolation,
  File "/Users/talley/Dropbox (HMS)/Python/forks/vispy/vispy/gloo/texture.py", line 133, in __init__
    self._resize(data.shape, format, internalformat)
  File "/Users/talley/Dropbox (HMS)/Python/forks/vispy/vispy/gloo/texture.py", line 300, in _resize
    internalformat = self._check_internalformat_change(internalformat, shape[-1])
  File "/Users/talley/Dropbox (HMS)/Python/forks/vispy/vispy/gloo/texture.py", line 288, in _check_internalformat_change
    raise ValueError('Internalformat does not match with given shape.')
ValueError: Internalformat does not match with given shape.
```

I tried to find a way to remove these two keys from this dict here, and put all the logic in the ComplexImageVisual, but couldn't figure it out without getting down to these `_scalable_texture` methods.  Let me know if this is a blocker.

Reply:
I just had a realization/idea about this. Are these errors regarding shape changes and stuff all during `__init__` when no initial data is provided to the parent `ImageVisual`? Would it be possible to pass some representative data along with forcing the internalformat to `r32f`/`rg32f`?